provider/adapters/src/fork/by_error.rs

```diff
@@ -0,0 +1,263 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::marker::PhantomData;
+
+use super::ForkByErrorPredicate;
+use alloc::vec::Vec;
+#[cfg(feature = "datagen")]
+use icu_provider::datagen;
+use icu_provider::prelude::*;
+
+/// A provider that returns data from one of two child providers based on a predicate function.
+///
+/// This is an abstract forking provider that must be provided with a type implementing the
+/// [`ForkByErrorPredicate`] trait.
+#[derive(Debug, PartialEq, Eq)]
+pub struct ForkByErrorProvider<P0, P1, F>(pub P0, pub P1, PhantomData<F>);
```

Comment:
thought: won't there be use cases for people to have a predicate that contains some data? I feel like we should use F here directly and have _most_ Fs be zero-sized types, PhantomData is typically for cases where you know there will be non-ZSTs that you wish to use as zero sized.

Reply:
If I did this, then I couldn't do my little trick where `ForkByKeyProvider` is just a typedef. But, maybe that's not a very clean approach anyway. I lean toward allowing the functions to have state, and `ForkByKeyProvider::new` can change to `ForkByErrorProvider::by_key`

Reply:
You could also have something like:

 - `FBEP::new()` which works for `F: Default`
 - `FBEP::with_predicate()` which works for all `F`

since it is easy for all ZSTs to implement `Default`

Reply:
but the naming change you proposed also works

Reply:
I renamed `FBEP::new` to `FBEF::new_with_predicate` and then implemented `FBKP::new` manually. It works with no changes to the call sites, because I quite like how the call sites currently read.